CONTRIBUTING.md

```diff
@@ -0,0 +1,584 @@
+  gitlab](https://gitlab.com/ginkgo-project/ginkgo-public-ci/pipelines) and
+  contains additional checks like static analysis and test coverage.
+* Once a PR has been approved and the build has passed, one of the reviewers can
+  mark the PR as `READY TO MERGE`. At this point the creator/assignee of the PR
+  *needs to* verify that the branch is up to date with `develop` and rebase it
+  on `develop` if it is not.
+
+
+## Code style
+
+### Automatic code formatting
+
+Ginkgo uses [ClangFormat](https://clang.llvm.org/docs/ClangFormat.html)
+(executable is usually named `clang-format`) and a custom `.clang-format`
+configuration file (mostly based on ClangFormat's _Google_ style) to
+automatically format your code. __Make sure you have ClangFormat set up and
+running properly__ ( you should be able to run `make format` from Ginkgo's build
+directory) before committing anything that will end up in a pull request against
+`ginkgo-project/ginkgo` repository. In addition, you should __never__ modify the
+`.clang-format` configuration file shipped with Ginkgo. E.g. if ClangFormat has
+trouble reading this file on your system, you should install a newer version of
+ClangFormat, and avoid commenting out parts of the configuration file.
+
+ClangFormat is the primary tool that helps us achieve a uniform look of Ginkgo's
+codebase, while reducing the learning curve of potential contributors. However,
+ClangFormat configuration is not expressive enough to incorporate the entire
+coding style, so there are several additional rules that all contributed code
+should follow.
+
+_Note_: To learn more about how ClangFormat will format your code, see existing
+files in Ginkgo, `.clang-format` configuration file shipped with Ginkgo, and
+ClangFormat's documentation.
+
+### Naming scheme
+
+#### Filenames 
+
+Filenames use `snake_case` and use the following extensions:
+*   C++ source files: `.cpp`
+*   C++ header files: `.hpp`
+*   CUDA source files: `.cu`
+*   CUDA header files: `.cuh`
+*   HIP source files: `.hip.cpp`
+*   HIP header files: `.hip.hpp`
+*   Common source files used by both CUDA and HIP: `.hpp.inc`
+*   CMake utility files: `.cmake`
+*   Shell scripts: `.sh`
+
+_Note:_ A C++ source/header file is considered a `CUDA` file if it contains CUDA
+code that is not guarded with `#if` guards that disable this code in non-CUDA
+compilers. I.e. if a file can be compiled by a general C++ compiler, it is not
+considered a CUDA file.
+
+#### Macros
+
+Macros (both object-like and function-like macros) use `CAPITAL_CASE`. They
+have to start with `GKO_` to avoid name clashes (even if they are `#undef`-ed in
+the same file!).
+
+#### Variables
+
+Variables use `snake_case`.
+
+#### Constants
+
+Constants use `snake_case`.
+
+#### Functions
+
+Functions use `snake_case`.
+
+#### Structures and classes
+
+Structures and classes which do not experience polymorphic behaviour (i.e. do
+not contain virtual methods, nor members which experience polymorphic behaviour)
```

Comment:
If the plan is to use American English throughout this document, then:
/s/behaviour/behavior